Library/Homebrew/cmd/leaves.rb

```diff
@@ -15,14 +15,40 @@ def leaves_args
       description <<~EOS
         List installed formulae that are not dependencies of another installed formula.
       EOS
+      switch "-r", "--installed-on-request",
+             description: "Only list leaves that were manually installed."
+      switch "-p", "--installed-as-dependency",
+             description: "Only list leaves that were installed as dependencies."
 
       named_args :none
     end
   end
 
+  def installed_on_request?(leaf)
+    Tab.for_keg(leaf.any_installed_keg).installed_on_request
+  end
+
+  def installed_as_dependency?(leaf)
+    Tab.for_keg(leaf.any_installed_keg).installed_as_dependency
+  end
+
   def leaves
-    leaves_args.parse
+    args = leaves_args.parse
+
+    leaves_list = Formula.installed_formulae_with_no_dependents
+
+    if args.installed_on_request?
+      leaves_list.select! do |l|
+        installed_on_request?(l)
+      end
+    end
+
+    if args.installed_as_dependency?
+      leaves_list.select! do |l|
+        installed_as_dependency?(l)
+      end
+    end
```

Comment:
```suggestion
    leaves_list.select!(&:installed_on_request?) if args.installed_on_request?
    leaves_list.select!(&:installed_as_dependency?) if args.installed_as_dependency?
```
Doesn't this work?

Also, you should probably make the flags mutually exclusive, as I believe passing both will return an empty list.

Reply:
> Doesn't this work?

Yes, thanks for the edit. Still getting used to writing idiomatic Ruby code.

> I believe passing both will return an empty list.

You're correct.

> Also, you should probably make the flags mutually exclusive

I considered doing this. In my mind, each flag applies a filter. I can't think of any reason that one would apply both filters, but the command behaves as expected when both filter flags are given. If you still think that the flags should be mutually exclusive, I'm happy to make the change.

Reply:
It's expected if you understand the internals of how these flags are used to construct lists of formulae. If you don't, one might expect passing both flags to list all the leaves. Of course, you could do that by omitting the flags entirely, but perhaps you are scripting these calls and think it's easier to pass both than to pass neither.

Alternatively, if you'd rather not make them mutually exclusive, I'd rather make passing both a no-op than return a blank list. We have to assume someone passing both flags is trying to get useful information out of `brew`, in which case just returning all leaves seems much more sensible here than return none.

Reply:
If I'm not wrong, `&:` would work only if there was a `leaf.installed_on_request?` method. This fails locally for me with:
```
➜ brew leaves --installed-on-request
Error: undefined method `installed_on_request?' for #<Formulary::FormulaNamespace04793c512083f42e82fc7403d806ad58::Dgraph:0x00007fac119b14e0>
Please report this issue:
  https://docs.brew.sh/Troubleshooting
/usr/local/Homebrew/Library/Homebrew/cmd/leaves.rb:40:in `select!'
/usr/local/Homebrew/Library/Homebrew/cmd/leaves.rb:40:in `leaves'
/usr/local/Homebrew/Library/Homebrew/brew.rb:122:in `<main>'
```
(and similarly for `--installed-as-dependency`).

> If you still think that the flags should be mutually exclusive, I'm happy to make the change.

Although one would probably not use them both together, I think we can make them mutually exclusive – that way it's more explicit.

Reply:
> If I'm not wrong, &: would work only if there was a leaf.installed_on_request? method (where leaf is an element in leaves_list). 

Yep, I forgot about that. If the operand of `&` isn't a `Proc`, Ruby calls `to_proc` on it. `Symbol#to_proc` returns a proc object that responds to the method given by the symbol on the receiver.

Reply:
oops, my mistake

Reply:
>  If you don't, one might expect passing both flags to list all the leaves.

Good point. I think making them mutually exclusive is the best bet here. Does the CLI parser have any built-in way of indicating that flags are mutually exclusive?

Reply:
You want to take a look at `conflicts`, like in:
https://github.com/Homebrew/brew/blob/a92c194bfbd6763e4f8a0d02f1135c0f646b5aa5/Library/Homebrew/cmd/list.rb#L57

Reply:
Thanks for the pointer!